code/modules/wiremod/duplicator.dm

```diff
@@ -0,0 +1,203 @@
+#define LOG_ERROR(list, error) if(list) { list.Add(error) }
+
+// Determines if a port can have a predefined input value if it is of this type.
+GLOBAL_LIST_INIT(circuit_dupe_whitelisted_types, list(
+	PORT_TYPE_NUMBER,
+	PORT_TYPE_STRING,
+	PORT_TYPE_ANY,
+))
+
+/// Loads a circuit based on json data at a location. Can also load usb connections, such as arrest consoles.
+/obj/item/integrated_circuit/proc/load_circuit_data(json_data, list/errors)
+	var/list/general_data = json_decode(json_data)
+
+	if(!general_data)
+		LOG_ERROR(errors, "Invalid json format!")
+		return
+
+	if(general_data["display_name"])
+		set_display_name(general_data["display_name"])
+
+	var/list/circuit_data = general_data["components"]
+	var/list/identifiers_to_circuit = list()
+	for(var/identifier in circuit_data)
+		var/list/component_data = circuit_data[identifier]
+		var/type = text2path(component_data["type"])
+		if(!ispath(type, /obj/item/circuit_component))
+			LOG_ERROR(errors, "Invalid path for circuit component, expected [/obj/item/circuit_component], got [type]")
```

Comment:
What's with the formatting here, on `[/obj/item/circuit_component]`?

Reply:
it was basically a way to make sure that if this type somehow doesn't get replaced if the type gets removed, it throws an error. I can change it to just be a string but I thought it'd be more sane to do it like this.

Reply:
Makes sense.